beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -306,55 +306,120 @@ func (s *Service) sendBatchRootRequest(ctx context.Context, roots [][32]byte, ra
 	ctx, span := trace.StartSpan(ctx, "sendBatchRootRequest")
 	defer span.End()
 
+	// Exit early if there are no roots to request.
+	if len(roots) == 0 {
+		return nil
+	}
+
+	// Remove duplicates (if any) from the list of roots.
 	roots = dedupRoots(roots)
-	s.pendingQueueLock.RLock()
-	for i := len(roots) - 1; i >= 0; i-- {
-		r := roots[i]
-		if s.seenPendingBlocks[r] || s.cfg.chain.BlockBeingSynced(r) {
-			roots = append(roots[:i], roots[i+1:]...)
-		} else {
+
+	// Reversly iterate through the list of roots to request blocks, and filter out roots that are already
+	// seen in pending blocks or being synced.
+	func() {
```

Comment:
having it as a separate `func()` makes it less readable. Can we have it as how it was before ?

Reply:
The purpose of this is to avoid any mutex `unlock` which is not deferred, as indicated in [this guideline](https://github.com/uber-go/guide/blob/master/style.md#defer-to-clean-up).

In the past, we had an issue with a `Rlock` and then an `Unlock` (without the `R), which cost us a multiple days debug time.

Having

```go
mut.Lock()
defer mut.Unlock()
```

really eases:
- To not forgot to unlock
- To unlock with the correct function